lib/shopify/enumerable.ex

```diff
@@ -0,0 +1,44 @@
+defmodule Shopify.Enumerable do
+  @moduledoc false
+
+  alias Shopify.{Error, Pagination, Response}
+
+  @enforce_keys [:data, :func, :middleware, :params, :session]
+  defstruct @enforce_keys
+
+  defimpl Enumerable do
+    # See https://hexdocs.pm/elixir/Enumerable.html
+
+    def count(_), do: {:error, __MODULE__}
+    def member?(_, _), do: {:error, __MODULE__}
+    def slice(_), do: {:error, __MODULE__}
+
+    def reduce(%Shopify.Enumerable{}, {:halt, acc}, _fun), do: {:halted, acc}
+
+    def reduce(%Shopify.Enumerable{} = enum, {:suspend, acc}, fun) do
+      {:suspended, acc, &reduce(enum, &1, fun)}
+    end
+
+    def reduce(%Shopify.Enumerable{data: [], params: nil}, {:cont, acc}, _fun) do
+      {:done, acc}
+    end
+
+    def reduce(%Shopify.Enumerable{data: []} = enum, {:cont, acc}, fun) do
+      case enum.middleware.(fn -> enum.func.(enum.session, enum.params) end) do
+        {:ok, %Response{data: data} = resp} ->
+          reduce(
+            %{enum | data: data, params: Pagination.next_page_params(resp)},
+            {:cont, acc},
+            fun
+          )
+
+        {:error, %Error{} = error} ->
```

Comment:
Here is a 3rd case that needs to be considered: 

`{:error, %Response{}}` e.g. Shopify fails (as the API does this relatively often) with a 500. Perhaps a retry in this case would be the best thing to do as paginating over a large collection of items can abort with high cost to re-iterate the whole collection.

Reply:
Thanks, I'll look into this. I have a busy week ahead of me so there's a chance I won't get to it until Monday.